wcfsetup/install/files/lib/data/user/UserProfileAction.class.php

```diff
@@ -567,6 +572,82 @@ public function updateSpecialTrophies()
         }
     }
 
+    /**
+     * Sets an avatar for a given user. The given file will be renamed and is gone after this method call.
+     *
+     * @throws UserInputException If none or more than one user is given.
+     * @throws \InvalidArgumentException If the given avatar is not an image or is incorrect sized.
+     */
+    public function setAvatar(): array
+    {
+        $user = $this->getSingleObject();
+
+        $imageData = \getimagesize($this->parameters['fileLocation']);
+
+        if (!$imageData) {
+            throw new \InvalidArgumentException("The given file is not an image.");
+        }
+
+        if ($imageData[0] != UserAvatar::AVATAR_SIZE || $imageData[1] != UserAvatar::AVATAR_SIZE) {
+            throw new \InvalidArgumentException(
+                "The given avatar does not have the size of " . UserAvatar::AVATAR_SIZE . "x" . UserAvatar::AVATAR_SIZE . "."
+            );
+        }
+
+        $data = [
+            'avatarName' => $this->parameters['filename'] ?? \basename($this->parameters['fileLocation']),
+            'avatarExtension' => ImageUtil::getExtensionByMimeType($imageData['mime']),
+            'width' => $imageData[0],
+            'height' => $imageData[1],
+            'userID' => $user->userID,
+            'fileHash' => \sha1_file($this->parameters['fileLocation']),
+        ];
+
+        // create avatar
+        $avatar = UserAvatarEditor::create($data);
+
+        try {
+            // check avatar directory
+            // and create subdirectory if necessary
+            $dir = \dirname($avatar->getLocation(null, false));
+            if (!\file_exists($dir)) {
+                FileUtil::makePath($dir);
+            }
+
+            \rename($this->parameters['fileLocation'], $avatar->getLocation(null, false));
+
+            // Create the WebP variant or the JPEG fallback of the avatar.
+            $avatarEditor = new UserAvatarEditor($avatar);
+            if ($avatarEditor->createAvatarVariant()) {
+                $avatar = new UserAvatar($avatar->avatarID);
+            }
+        } catch (\Exception $e) {
+            $editor = new UserAvatarEditor($avatar);
+            $editor->delete();
+
+            throw $e;
+        }
+
+        // delete old avatar
+        if ($user->avatarID) {
+            (new UserAvatarAction([$user->avatarID], 'delete'))->executeAction();
+        }
+
+        // update user
+        $userEditor = new UserEditor($user->getDecoratedObject());
+        $userEditor->update([
+            'avatarID' => $avatar->avatarID,
+            'enableGravatar' => 0,
+        ]);
```

Comment:
Should this also be part of the `try`?

Reply:
Maybe the code from 636 to 641 would make sense there. The deletion of the old avatar imo not.